packages/contract-utils/upgradeable/src/upgradeable.rs

```diff
@@ -0,0 +1,112 @@
+use soroban_sdk::{contractclient, contracterror, Address, BytesN, Env, FromVal, Val};
+
+/// High-level trait for contract upgrades.
+///
+/// This trait defines the external entry point and can be used in two ways:
+///
+/// 1. Standalone – Implement this trait directly when full control over access
+///    control and upgrade logic is required. In this case, the implementor is
+///    responsible for ensuring:
+///    - Proper authorization of the `operator`
+///    - Versioning management
+///
+/// 2. Framework-assisted usage – When using the lightweight upgrade framework
+///    provided in this module, you should NOT manually implement this trait.
+///    Instead:
+///    - Derive it using `#[derive(Upgradeable)]`
+///    - Provide access control by implementing [`UpgradeableInternal`] with
+///      your custom logic
+#[contractclient(name = "UpgradeableClient")]
+pub trait Upgradeable {
```

Comment:
should this have a bound `Upgradable: UpgradableInternal`? 
I believe, you didn't put it because someone may implement it in a custom way that wouldn't require `UpgradableInternal`. 

But couldn't we use the same argument for `Migratable: MigratableInternal`?

Reply:
Yes, that's correct about `UpgradableInternal`, but `Migratable` and `MigratableInternal` are interconnected as they share the same accoc types. If someone wants something that doesn't fit within the provided framework, they simply don't need those traits.